keps/sig-apps/4563-evacuation-api/README.md

```diff
@@ -0,0 +1,1561 @@
+- Easily identify pods that have accepted evacuation and are making progress. If possible to be
+  able to see evacuation's ETA.
+- Identify pods that should be evicted instead of evacuated and to distinguish pods that are
+  failing eviction.
+- See additional debug information from the active evacuator and be able to identify all the
+  registered evacuators.
+
+#### Story 2
+
+As an application owner, I want to run single replica applications without disruptions and have the
+ability to easily migrate the workload pods from one node to another. This also applies to
+applications with larger number of replicas that prefer to surge (upscale) pods first instead of
+downscaling.
+
+#### Story 3
+
+As an application owner, I want to have a custom logic tailored to my application for migration,
+down-scaling, or removal of pods. I want to be able to easily override the default evacuation
+process, including the eviction and PDBs, available to workloads. To achieve this, I also need to
+be able to identify other evacuators and an order in which they will run.
+
+#### Story 4
+
+As an application owner I want my pods to be scheduled on correct nodes. I want to use the
+descheduler or the upcoming Affinity Based Eviction feature to remove pods from incorrect nodes
+and then have the pods scheduled on new ones. I want to do the rescheduling gracefully and be able
+to control the disruption level of my application (even 0% application unavailability).
+
+### Notes/Constraints/Caveats (Optional)
+
+#### Length Limitations for Pod Annotations and Evacuation Finalizers
+
+We use the following finalizers in Evacuation objects:
+`evacuation.coordination.k8s.io/instigator_${EVACUATION_INSTIGATOR_SUBDOMAIN}`.
+
+We use the following annotations in Pod objects:
+`evacuation.coordination.k8s.io/priority_${EVACUATOR_CLASS}: ${PRIORITY}/${ROLE}`.
+
+Key validation for both of these is subject to [IsQualifiedName](https://github.com/kubernetes/kubernetes/blob/cae35dba5a3060711a2a3f958537003bc74a59c0/staging/src/k8s.io/apimachinery/pkg/util/validation/validation.go#L42)
+validation. The key consists of `prefix/name`. The maximum length of the name is 63 characters,
+which may not always be enough for `EVACUATION_INSTIGATOR_SUBDOMAIN` and `EVACUATOR_CLASS` (also
+DNS subdomain) after adding the length of `instigator_` and `priority_` respectively.
+
+If the name component of the key is greater than 63 characters, it is recommended to use a unique
+shortcut of your subdomain name.
+For example, `deployment.apps.k8s.io` instead of `deployment.apps.kubernetes.io`.
+
+We do not expect a large number of instigators and evacuators. If this becomes a larger problem in
+the future, we can implement an additional scheme:
+
+```yaml
+annotations:
+  evacuation.coordination.k8s.io/priority_${EVACUATOR_CLASS_HASH}: ${PRIORITY}/${ROLE}
+  evacuation.coordination.k8s.io/hash_${EVACUATOR_CLASS_HASH}: ${EVACUATOR_CLASS}
+```
+
+which would result in
+
+```yaml
+annotations:
```

Comment:
Why is annotation the mechanism for this?   Why not a "real" API with fields?  I'm not a fan of "stringly typed" data and embedding data into the NAME of an annotation is something we try not to do.

Reply:
I think it is quite hard to achieve this for pods.
- I think it is preferable not to change the `.spec` of the pod once it is running. It does not seem like a good place to store transient values (evacuators can change during the lifetime of a pod).
- It does not seem like a good match for the `.status` either. We expect multiple actors to reconcile these values, so giving them RBAC access to the pod status might end up corrupting the status. I am also not sure if the status reconciliation would play well with kubelet.